### PR TITLE
Write Power Transformer functions for xarray compatibility

### PR DESCRIPTION
**Pull request by veni-vidi-vici-dormivi**
Here I write functions for the power transformer to take in xarrays.

 - [ ] Tests added
 - [ ] Fully documented, including `CHANGELOG.rst`

### CODE REVIEW

**Comment by veni-vidi-vici-dormivi** [comment]
At the moment I am appending functions for the xarray power transformer below the old functions. I also now opted to write functions instead of class methods. This is WIP though we can build on this first draft here.

File: mesmer/mesmer_m/power_transformer.py [hunk 1]
```diff
@@ -290,3 +291,233 @@ def inverse_transform(self, transformed_monthly_T, yearly_T):
...
+            _yeo_johnson_optimize_lambda,
+            monthly_residuals_grouped[month],
+            yearly_T,
+            input_core_dims=[["time"], ["time"]],
+            output_core_dims=[[], []],
+            output_dtypes=[float, float],
+            vectorize=True,
+            join="outer",
+        )
+        data = {"xi_0": xi_0, "xi_1": xi_1}
```
Comment (mathause) [inline mesmer/mesmer_m/power_transformer.py:449]:
```suggestion
        data_vars = {"xi_0": xi_0, "xi_1": xi_1}
```

(or create the `Dataset` directly)

File: mesmer/mesmer_m/power_transformer.py [hunk 2]
```diff
@@ -290,3 +291,233 @@ def inverse_transform(self, transformed_monthly_T, yearly_T):
...
+
+def _get_lambdas_from_covariates_xr(coeffs, yearly_T):
+
+    lambdas = 2 / (1 + coeffs.xi_0 * np.exp(yearly_T * coeffs.xi_1))
+
+    return lambdas.rename("lambdas")
+
+
+def fit_power_transformer_xr(monthly_residuals, yearly_T):
+    """Estimate the optimal parameter lambda for each gridcell, given an xarray dataset
```
Comment (mathause) [inline mesmer/mesmer_m/power_transformer.py:416]: Careful using `dataset` as a generic name

**Review by mathause — COMMENTED** [review]
- Can you fix the docstrings and comments so they no wider than the 88 characters.
- We need to align the names of the main functions - not sure if I prefer `power_transformer` or `yeo_johnson`... On second thought I think `yeo_johnson` because `power_transformer` includes the box-cox transformation. Currently they are not consistent:
  ```python
  power_transformer.fit_power_transformer_xr(...)
  power_transformer.yeo_johnson_transform_xr(...)
  power_transformer.inverse_transform(...)
  ```
- Can still consider what gives the better naming & usability - 3 functions or a class (no worries this would not mean much additional work, as the functions are `@staticmethod`)
- What is your suggestion for the next steps i.e. getting rid of the `PowerTransformerVariableLambda` class? (I thought about advising to rip out the internals and replace them with your new functions, but I am less sure if we remove the whole class next)

**Comment by veni-vidi-vici-dormivi** [comment]
> Can you fix the docstrings and comments so they no wider than the 88 characters.

Does the precommit hook not work for docstrings and comments?

**Comment by mathause** [comment]
Yes I am no longer happy about the approach I took for the linear regression - especially the  `lr.params = params` part. Is creating a class just a cheap way around the flatter hierarchy we discussed in #348 (i.e. having everything in the `mesmer.stats` namespace)? But then again - having 3-4 functions for each method makes it cluttered fast... Really not sure what is best... we may have to re-visit #348 at some point...

Sticking with functions for now, maybe?

```python
mesmer.stats.yeo_johnson_fit(...)
mesmer.stats.yeo_johnson_transform(...)
mesmer.stats.yeo_johnson_inverse_transform(...)
```

File: tests/unit/test_power_transformer.py [hunk 3]
```diff
@@ -151,3 +154,92 @@ def test_standard_scaler():
...
+    pt_coefficients = power_transformer.fit_yeo_johnson_transform(
+        monthly_residuals, yearly_T
+    )
+    transformed = power_transformer.yeo_johnson_transform_xr(
+        monthly_residuals, pt_coefficients, yearly_T
+    )
+    inverse_transformed = power_transformer.inverse_yeo_johnson_transform_xr(
+        transformed.transformed, pt_coefficients, yearly_T
+    )
+    xr.testing.assert_allclose(inverse_transformed.inverted, monthly_residuals)
```
Comment (mathause) [inline tests/unit/test_power_transformer.py:205]: Add checks to ensure the shape and size is correct?

**Review by mathause — APPROVED** [review]
Thanks - I think that's basically ready.

One thing we need to think about at one point - how do we best store coeffs in a Dataset/ DataArray - is it better to have individual arrays (as here or for lin reg) or an array of params? (It probably depends on the kind of coeffs...)

---


Can you run velin (a docstring checker) on the file?

```bash
pip install velin

velin mesmer/mesmer_m/power_transformer.py
# velin mesmer/mesmer_m/power_transformer.py --write
```

(Can also run it on the full project, but prob. in another PR, also `velin .` will throw an error because it picks up the .git folder, so need to do `velin mesmer`.)

**Pull request merged**